nni/common/serializer.py

```diff
@@ -0,0 +1,369 @@
+    class wrapper(SerializableObject, base):
+        def __init__(self, *args, **kwargs):
+            # store a copy of initial parameters
+            args, kwargs = _get_arguments_as_dict(base.__init__, args, kwargs, kw_only)
+
+            # calling serializable object init to initialize the full object
+            super().__init__(symbol=base, args=args, kwargs=kwargs, _self_contained=True)
+
+    _MISSING = '_missing'
+    for k in functools.WRAPPER_ASSIGNMENTS:
+        # assign magic attributes like __module__, __qualname__, __doc__
+        v = getattr(base, k, _MISSING)
+        if v is not _MISSING:
+            try:
+                setattr(wrapper, k, v)
+            except AttributeError:
+                pass
+
+    return wrapper
+
+
+def _trace_func(func, kw_only):
+    @functools.wraps
+    def wrapper(*args, **kwargs):
+        # similar to class, store parameters here
+        args, kwargs = _get_arguments_as_dict(func, args, kwargs, kw_only)
+        return SerializableObject(func, args, kwargs)
+
+    return wrapper
+
+
+def _get_arguments_as_dict(func, args, kwargs, kw_only):
+    if kw_only:
+        # get arguments passed to a function, and save it as a dict
+        argname_list = list(inspect.signature(func).parameters.keys())[1:]
+        full_args = {}
+        full_args.update(kwargs)
+
+        # match arguments with given arguments
+        # args should be longer than given list, because args can be used in a kwargs way
+        assert len(args) <= len(argname_list), f'Length of {args} is greater than length of {argname_list}.'
+        for argname, value in zip(argname_list, args):
+            full_args[argname] = value
+
+        args, kwargs = [], full_args
+
+    return args, kwargs
+
+
+def _import_cls_or_func_from_name(target: str) -> Any:
+    if target is None:
+        return None
+    path, identifier = target.rsplit('.', 1)
+    module = __import__(path, globals(), locals(), [identifier])
+    return getattr(module, identifier)
+
+
+def _get_cls_or_func_name(cls_or_func: Any) -> str:
+    module_name = cls_or_func.__module__
+    if module_name == '__main__':
+        raise ImportError('Cannot use a path to identify something from __main__.')
+    full_name = module_name + '.' + cls_or_func.__name__
+
+    try:
+        imported = _import_cls_or_func_from_name(full_name)
+        if imported != cls_or_func:
+            raise ImportError(f'Imported {imported} is not same as expected. The function might be dynamically created.')
+    except ImportError:
+        raise ImportError(f'Import {cls_or_func.__name__} from "{module_name}" failed.')
+
+    return full_name
+
+
+def _get_hybrid_cls_or_func_name(cls_or_func: Any, pickle_size_limit: int = 4096) -> str:
+    try:
+        name = _get_cls_or_func_name(cls_or_func)
+        # import success, use a path format
+        return 'path:' + name
+    except ImportError:
+        b = cloudpickle.dumps(cls_or_func)
+        if len(b) > pickle_size_limit:
```

Comment:
how to avoid serializing data in dataloader?

Reply:
You can serialize a factory of dataloader.